client-server-client/src/main/scala/zio/raft/client/RaftClient.scala

```diff
@@ -0,0 +1,702 @@
+                  )
+                  .as(this)
+            }
+
+          case StreamEvent.ServerMsg(RequestError(NotLeaderRequest, leaderId)) =>
+            for {
+              currentAddr = addresses(currentAddressIndex)
+              _ <- ZIO.logInfo(s"Not leader at $currentAddr, reconnecting")
+              nonce <- Nonce.generate()
+              nextIndex = (currentAddressIndex + 1) % addresses.length
+              nextAddr = addresses(nextIndex)
+              _ <- transport.disconnect().orDie
+              _ <- transport.connect(nextAddr).orDie
+              _ <- transport.sendMessage(ContinueSession(sessionId, nonce)).orDie
+              _ <- ZIO.logInfo(s"Connecting Existing Session $sessionId to $nextAddr")
+              now <- Clock.instant
+            } yield ConnectingExistingSession(
+              sessionId = sessionId,
+              capabilities = capabilities,
+              nonce = nonce,
+              addresses = addresses,
+              currentAddressIndex = nextIndex,
+              createdAt = now,
+              serverRequestTracker = serverRequestTracker,
+              nextRequestId = nextRequestId,
+              pendingRequests = pendingRequests
+            )
+
+          case StreamEvent.ServerMsg(RequestError(SessionTerminated, _)) =>
+            ZIO.dieMessage(s"Session terminated by server for session $sessionId")
+
+          case StreamEvent.ServerMsg(RequestError(reason, _)) =>
+            ZIO.logWarning(s"Request error: $reason").as(this)
+
+          case StreamEvent.ServerMsg(SessionClosed(reason, _)) =>
+            for {
+              _ <- ZIO.logInfo(s"Session closed: $reason, reconnecting")
+              nonce <- Nonce.generate()
+              currentAddr = addresses(currentAddressIndex)
+              _ <- transport.disconnect().orDie
+              _ <- transport.connect(currentAddr).orDie
+              _ <- transport.sendMessage(ContinueSession(sessionId, nonce)).orDie
+              now <- Clock.instant
+            } yield ConnectingExistingSession(
+              sessionId = sessionId,
+              capabilities = capabilities,
+              nonce = nonce,
+              addresses = addresses,
+              currentAddressIndex = currentAddressIndex,
+              createdAt = now,
+              serverRequestTracker = serverRequestTracker,
+              nextRequestId = nextRequestId,
+              pendingRequests = pendingRequests
+            )
+
+          case StreamEvent.KeepAliveTick =>
+            for {
+              now <- Clock.instant
+              _ <- transport.sendMessage(KeepAlive(now)).orDie
+            } yield this
+
+          case StreamEvent.TimeoutCheck =>
+            for {
+              now <- Clock.instant
+              newPending <- pendingRequests.resendExpired(transport, now, ClientConfig.REQUEST_TIMEOUT)
```

Comment:
REQUEST_TIMEOUT is the default value, take the one from clientConfig.RequestTimeout